geomstats/learning/kmeans.py

```diff
@@ -103,24 +105,43 @@ def fit(self, X):
         n_samples = X.shape[0]
         if self.verbose > 0:
             logging.info("Initializing...")
-        if self.init == "kmeans++":
-            centroids = [gs.expand_dims(X[randint(0, n_samples - 1)], 0)]
-            for i in range(self.n_clusters - 1):
-                dists = [
-                    gs.to_ndarray(self.metric.dist(centroids[j], X), 2, 1)
-                    for j in range(i + 1)
+
+        if isinstance(self.init, str):
+            if self.init == "kmeans++":
+                centroids = [gs.expand_dims(X[randint(0, n_samples - 1)], 0)]
+                for i in range(self.n_clusters - 1):
+                    dists = [
+                        gs.to_ndarray(self.metric.dist(centroids[j], X), 2, 1)
+                        for j in range(i + 1)
+                    ]
+                    dists = gs.hstack(dists)
+                    dists_to_closest_centroid = gs.amin(dists, 1)
+                    indices = gs.arange(n_samples)
+                    weights = dists_to_closest_centroid / gs.sum(
+                        dists_to_closest_centroid
+                    )
+                    index = rv_discrete(values=(indices, weights)).rvs()
+                    centroids.append(gs.expand_dims(X[index], 0))
+            elif self.init == "random":
+                centroids = [
+                    gs.expand_dims(X[randint(0, n_samples - 1)], 0)
+                    for i in range(self.n_clusters)
                 ]
-                dists = gs.hstack(dists)
-                dists_to_closest_centroid = gs.amin(dists, 1)
-                indices = gs.arange(n_samples)
-                weights = dists_to_closest_centroid / gs.sum(dists_to_closest_centroid)
-                index = rv_discrete(values=(indices, weights)).rvs()
-                centroids.append(gs.expand_dims(X[index], 0))
+            else:
+                raise ValueError(f"Unknown initial centroids method '{self.init}'.")
         else:
+            if self.init.shape[0] != self.n_clusters:
+                raise ValueError("Need as many initial centroids as clusters.")
+
+            if self.init.shape[1] != X.shape[1]:
+                raise ValueError(
+                    "Dimensions of initial centroids and training data do not match."
+                )
+
             centroids = [
-                gs.expand_dims(X[randint(0, n_samples - 1)], 0)
-                for i in range(self.n_clusters)
+                gs.expand_dims(self.init[i], 0) for i in range(self.n_clusters)
             ]
+
         self.centroids = gs.concatenate(centroids, axis=0)
```

Comment:
Here I think we can do:

```
self.centroids = gs.copy(centroids)
self.init_centroids = gs.copy(centroids)
```

And above, within the `if` branch that handles `str`, do: `centroids = gs.concatenate(centroids, axis=0)`.

So we simplify the new if branch by simply doing: `centroids = init` (no need for `gs.expand_dims`).

Reply:
Makes sense. I also implemented the `callable` option. Note that my approach requires Python >= 3.2 as I'm using the `callable()` built-in function.

Reply:
Thanks! We assume Python >= 3.7, so it is perfectly fine.